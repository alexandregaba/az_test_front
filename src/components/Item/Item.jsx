import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import CheckBox from '../CheckBox/CheckBox';
import NameField from '../NameField/NameField';
import Button from '../Button/Button';

const Li = styled.li`
  list-style-type: none;
  width: 100%;
  border-bottom: solid 1px gainsboro;
  margin-bottom: 5px;
`;

const Div = styled.div`
  width: 100%;
  display: flex;
  flex-direction: ${props => (props.reverse ? 'row-reverse' : 'row')};
`;

class Item extends Component {
  constructor(props) {
    super(props);
    this.handleEdit = this.handleEdit.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleDeleteItem = this.handleDeleteItem.bind(this);
    this.handleCheck = this.handleCheck.bind(this);
    this.state = { editInputValue: this.props.name };
  }

  static propTypes = {
    name: PropTypes.string,
    id: PropTypes.string,
    checked: PropTypes.bool,
    editMode: PropTypes.bool,
    startEditGroceryItem: PropTypes.func,
    stopEditGroceryItem: PropTypes.func,
  };

  static defaultProps = {
    name: '',
    id: 1,
    checked: false,
    editMode: false,
    startEditGroceryItem: () => {},
    stopEditGroceryItem: () => {},
  };

  handleEdit() {
    if (!this.props.editMode) {
      this.props.startEditGroceryItem(this.props.id);
    } else {
      const { checked, id } = this.props;
      const item = {
        name: this.state.editInputValue,
        checked,
        id,
      };
      this.props.stopEditGroceryItem(item);
    }
  }

  onPressEnter = event => {
    if (event.key === 'Enter') {
      this.handleEdit();
    }
  };

  handleInputChange(event) {
    this.setState({ editInputValue: event.target.value });
  }

  handleDeleteItem() {
    this.props.deleteGroceryItem(this.props.id);
  }

  handleCheck() {
    this.props.toggleCheckGroceryItem(this.props.id);
  }

  render() {
    const { name, editMode, checked } = this.props;
    const { editInputValue } = this.state;
    return (
      <Li>
        <Div>
          <CheckBox checked={checked} onClick={this.handleCheck} />
          <NameField
            checked={checked}
            name={name}
            value={editInputValue}
            editMode={editMode}
            handleInputChange={this.handleInputChange}
            onPressEnter={this.onPressEnter}
          />
        </Div>
        <Div reverse>
          <Button type="delete" handleDelete={this.handleDeleteItem} editMode={editMode} />
          <Button type="edit" handleEdit={this.handleEdit} editMode={editMode} />
        </Div>
      </Li>
    );
  }
}
export default Item;
