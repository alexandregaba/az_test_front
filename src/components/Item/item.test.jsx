import React from 'react';
import Item from './Item';

import { shallow } from 'enzyme';

const props = {
  name: 'orange',
  id: '1',
  checked: false,
  editMode: false,
  startEditGroceryItem: jest.fn(),
  stopEditGroceryItem: jest.fn(),
  deleteGroceryItem: jest.fn(),
  toggleCheckGroceryItem: jest.fn(),
};

it('match snapshot', () => {
  const props = {
    checked: false,
    name: 'Apples',
    id: '1',
    editMode: false,
  };
  const Component = shallow(<Item {...props} />);
  expect(Component).toMatchSnapshot();
});

it('it triggers startEditGroceryItem when editButton is clicked', () => {
  const Component = shallow(<Item {...props} />);
  Component.find('#editButton').simulate('click');
  expect(props.startEditGroceryItem).toHaveBeenCalledWith('1');
});

it('it triggers stopEditGroceryItem when editButton is clicked in editMode', () => {
  const Component = shallow(<Item {...props} editMode />);
  Component.find('#editButton').simulate('click');
  expect(props.stopEditGroceryItem).toHaveBeenCalledWith({
    checked: false,
    id: '1',
    name: 'orange',
  });
});

it('it triggers deleteGroceryItem when deleteButton is clicked', () => {
  const Component = shallow(<Item {...props} editMode />);
  Component.find('#deleteButton').simulate('click');
  expect(props.deleteGroceryItem).toHaveBeenCalledWith('1');
});

it('it triggers toggleCheckGroceryItem when itemCheck value is changed', () => {
  const Component = shallow(<Item {...props} editMode />);
  Component.find('#itemCheck').simulate('change', { target: { checked: true } });
  expect(props.toggleCheckGroceryItem).toHaveBeenCalledWith('1');
});
