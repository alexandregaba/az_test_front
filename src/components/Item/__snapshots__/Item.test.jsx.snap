// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<styled.li>
  <styled.div>
    <CheckBox
      checked={false}
      onClick={[Function]}
    />
    <NameField
      checked={false}
      editMode={false}
      handleInputChange={[Function]}
      name="Apples"
      onPressEnter={[Function]}
      value="Apples"
    />
  </styled.div>
  <styled.div
    reverse={true}
  >
    <Button
      editMode={false}
      handleDelete={[Function]}
      type="delete"
    />
    <Button
      editMode={false}
      handleEdit={[Function]}
      type="edit"
    />
  </styled.div>
</styled.li>
`;
